controller/api/destination/endpoint_translator.go

```diff
@@ -287,6 +303,120 @@ func (et *endpointTranslator) sendClientAdd(set watcher.AddressSet) {
 	}
 }
 
+// watchEndpointPolicy establishes a watch on the the policy server for each
+// pod address.
+//
+// The policy server client will error if the stream is closed or the pod
+// address is removed. When the pod address is removed, closeEndpointPolicy
+// will close the client.
+//
+// The policy server client handles Recv errors and signals the updates
+// routine to shutdown.
+func (et *endpointTranslator) watchEndpointPolicy(set watcher.AddressSet) {
+	for _, addr := range set.Addresses {
+		// If the address is not backed by a pod then there will be no policy
+		// server to watch.
+		if addr.Pod == nil {
+			continue
+		}
+
+		opaquePorts, ok, err := getPodOpaquePortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get opaque ports annotation for pod: %s", err)
+		}
+
+		// If the opaque ports annotation was not set, then use the
+		// endpoint translator's default value.
+		if !ok {
+			opaquePorts = et.defaultOpaquePorts
+		}
+
+		skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(addr.Pod)
+		if err != nil {
+			et.log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
+		}
+
+		// Create a client which watches for changes to the inbound policy for
+		// portSpec.
+		policyPortSpec := &policyPb.PortSpec{
+			Workload: fmt.Sprintf("%s:%s", addr.Pod.Namespace, addr.Pod.Name),
+			Port:     addr.Port,
+		}
+		client, err := et.policyClient.WatchPort(et.stream.Context(), policyPortSpec)
+		if err != nil {
+			et.log.Errorf("failed to create policy server watch for %s:%d: %s", policyPortSpec.Workload, policyPortSpec.Port, err)
+		}
+		et.log.Infof("Establishing watch on policy server %s:%d", policyPortSpec.Workload, policyPortSpec.Port)
+
+		// policyPb.PortSpec cannot be used as a map key, so we create an
+		// equivalent port spec with comparable keys.
+		portSpec := &portSpec{
+			workload: policyPortSpec.Workload,
+			port:     policyPortSpec.Port,
+		}
+		et.policyWatches[*portSpec] = client
```

Comment:
do we need to worry about if there is already a policy watch in the map for this portSpec?  as written, I think the previous watch would be overwritten in the map and thus be leaked and no longer be cancellable.  If we add checking to see if an entry already exists in the map, we probably also have to add some locking to avoid race conditions.

Reply:
I wrapped the policy watches with a new `policyWatches` struct that holds the watches and a mutex. `watchEndpointPolicy` and `closeEndpointPolicy` now both lock `policyWatches` before adding/deleting clients from the `watches` map.